natss/pkg/dispatcher/dispatcher.go

```diff
@@ -290,40 +276,38 @@ func (s *SubscriptionsSupervisor) UpdateSubscriptions(channel *messagingv1alpha1
 	return failedToSubscribe, nil
 }
 
-func (s *SubscriptionsSupervisor) subscribe(channel eventingchannels.ChannelReference, subscription subscriptionReference) (*stan.Subscription, error) {
+func (s *SubscriptionsSupervisor) subscribe(ctx context.Context, channel eventingchannels.ChannelReference, subscription subscriptionReference) (*stan.Subscription, error) {
 	s.logger.Info("Subscribe to channel:", zap.Any("channel", channel), zap.Any("subscription", subscription))
 
 	mcb := func(msg *stan.Msg) {
-		event := cloudevents.Event{}
-		err := event.UnmarshalJSON(msg.Data)
-		s.logger.Info("Received event", zap.String("channel", channel.String()))
+		event, err := natsscloudevents.NewMessage(msg, natsscloudevents.WithManualAcks())
 		if err != nil {
-			s.logger.Error(err.Error(), zap.Error(err))
+			s.logger.Error("could not create a message", zap.Error(err))
 			return
 		}
-		if err := event.Validate(); err != nil {
-			s.logger.Error(err.Error(), zap.Error(err))
-			return
-		}
-		s.logger.Sugar().Debugf("NATSS message received from subject: %v; sequence: %v; timestamp: %v, event: '%s'", msg.Subject, msg.Sequence, msg.Timestamp, event.String())
-		if err := s.dispatcher.DispatchEventWithDelivery(context.TODO(), event, subscription.SubscriberURI, subscription.ReplyURI, &subscription.Delivery); err != nil {
+		s.logger.Debug("NATSS message received", zap.String("subject", msg.Subject), zap.Uint64("sequence", msg.Sequence), zap.Time("timestamp", time.Unix(msg.Timestamp, 0)))
+		if err := s.dispatcher.DispatchMessageWithDelivery(ctx, event, nil, subscription.SubscriberURI, subscription.ReplyURI, &subscription.Delivery); err != nil {
 			s.logger.Error("Failed to dispatch message: ", zap.Error(err))
 			return
 		}
 		if err := msg.Ack(); err != nil {
-			s.logger.Error("Failed to acknowledge message: ", zap.Error(err))
+			s.logger.Error("failed to acknowledge message", zap.Error(err))
 		}
 	}
-	// subscribe to a NATSS subject
+
 	ch := getSubject(channel)
 	sub := subscription.String()
+
 	s.natssConnMux.Lock()
 	currentNatssConn := s.natssConn
```

Comment:
i have no clue how nats/stan works, but this lock looks weird to me. @n3wscott ?